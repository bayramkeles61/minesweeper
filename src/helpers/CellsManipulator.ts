import { Cell, Coords, Field } from "./Field";

/**
 * Get neighbour cells indexes
 * @param {Coords} coords
 * @returns {Record<string, [number, number]>}
 */
export const getNeighboursItems = ([y, x]: Coords): Record<
  string,
  [number, number]
> => ({
  top: [y - 1, x],
  topRight: [y - 1, x + 1],
  right: [y, x + 1],
  rightBottom: [y + 1, x + 1],
  bottom: [y + 1, x],
  bottomLeft: [y + 1, x - 1],
  left: [y, x - 1],
  leftTop: [y - 1, x - 1],
});

/**
 * Check item in the field
 * @param {Coords} coords
 * @param {Field} field
 * @returns {boolean}
 */
export const checkItemInField = ([y, x]: Coords, { length }: Field): boolean =>
  y >= 0 && x >= 0 && length - y > 0 && length - x > 0;

/**
 * Increment neighbour items for cell with coords
 * @param {Coords} coords
 * @param {Field} field
 * @returns {Cell}
 */
export const incrementNeighbours = (coords: Coords, field: Field): Field => {
  const items = getNeighboursItems(coords);

  for (const [y, x] of Object.values(items)) {
    if (checkItemInField([y, x], field)) {
      const cell = field[y][x];
      if (cell < 8) {
        field[y][x] = (cell + 1) as Cell;
      }
    }
  }

  return field;
};
