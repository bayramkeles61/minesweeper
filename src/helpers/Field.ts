import { incrementNeighbours } from "./CellsManipulator";

export type Cell = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12;
export type Field = Cell[][];
export type Coords = [number, number];

export const CellState: Record<string, Cell> = {
  empty: 0,
  bomb: 9,
  hidden: 10,
  flag: 11,
  weakFlag: 12,
};

/**
 * Create empty field
 *
 * Example emptyFieldGenerator(2)
 * [
 *   [0, 0],
 *   [0, 0],
 * ]
 *
 * @param {number} size
 * @param {Cell} state
 * @returns {Field}
 */
export const generateFieldWithDefaultState = (
  size: number,
  state: Cell = CellState.empty,
): Field => new Array(size).fill(null).map(() => new Array(size).fill(state));

/**
 * Generate field with mines
 * Cell has a range between 0 and 9 where
 * 0 - have no any mine
 * from 0 to 8 - count of mines arrount the cell
 * 9 - special flag for the mine
 *
 * (1/9 ~ 11.11%)
 * Example fieldGenerator(3, 1/9) would return
 * [
 *   [1, 1, 1],
 *   [1, 9, 1],
 *   [1, 1, 1]
 * ]
 *
 * @param {number} size
 * @param {number} probability
 * @returns {Field}
 */
export const fieldGenerator = (size: number, probability: number): Field => {
  if (probability < 0 || probability > 1) {
    throw new Error("Probability must be between 0 and 1");
  }

  let unprocessedCells = size * size;
  let restCellsWithBombs = unprocessedCells * probability;

  const result: Field = generateFieldWithDefaultState(size);

  // Stryker disable next-line EqualityOperator
  for (let y = 0; y < size; y++) {
    for (let x = 0; x < size; x++) {
      // Stryker disable next-line EqualityOperator
      if (restCellsWithBombs / unprocessedCells > Math.random()) {
        result[y][x] = CellState.bomb;
        incrementNeighbours([y, x], result);
        restCellsWithBombs--;
      }
      unprocessedCells--;
    }
  }
  return result;
};
