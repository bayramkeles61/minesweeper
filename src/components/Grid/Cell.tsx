import React, { FC, memo } from "react";
import styled from "@emotion/styled";

import { Cell as CellType, Coords, CellState } from "@/helpers/Field";

import { useMouseDown } from "@/hooks/useMouseDown";

export interface CellProps {
  /**
   * Cell status based on the CellType
   */
  children: CellType;
  /**
   * Cell coordinates
   */
  coords: Coords;
  /**
   * onClick by cell handler
   */
  onClick: (coords: Coords) => void;
  /**
   * onContextMenu by cell handler
   */
  onContextMenu: (coords: Coords) => void;
}

export const isActiveCell = (cell: CellType): boolean =>
  [CellState.hidden, CellState.flag, CellState.weakFlag].includes(cell);

export const areEqual = (
  prevProps: CellProps,
  nextProps: CellProps,
): boolean => {
  const areEqualCoords =
    prevProps.coords.filter((coord, idx) => nextProps.coords[idx] !== coord)
      .length === 0;

  return (
    prevProps.children === nextProps.children &&
    areEqualCoords &&
    prevProps.onClick === nextProps.onClick &&
    prevProps.onContextMenu === nextProps.onContextMenu
  );
};

export const Cell: FC<CellProps> = memo(({ children, coords, ...rest }) => {
  const [mouseDown, onMouseDown, onMouseUp] = useMouseDown();

  const onClick = () => rest.onClick(coords);

  const onContextMenu = (event: React.MouseEvent<HTMLDivElement>) => {
    /**
     * Prevent context menu by default
     */
    event.preventDefault();

    if (isActiveCell(children)) {
      rest.onContextMenu(coords);
    }
  };

  const props = {
    onClick,
    onContextMenu,
    onMouseDown,
    onMouseUp,
    onMouseLeave: onMouseUp,
    mouseDown,
    "data-testid": `${coords}`,
    role: "cell",
  };

  return <ComponentsMap {...props}>{children}</ComponentsMap>;
}, areEqual);

// Stryker disable next-line StringLiteral
Cell.displayName = "Cell";

interface ComponentsMapProps {
  children: CellType;
  onClick: (event: React.MouseEvent<HTMLDivElement>) => void;
  onContextMenu: (event: React.MouseEvent<HTMLDivElement>) => void;
  onMouseDown: () => void;
  onMouseUp: () => void;
  onMouseLeave: () => void;
  mouseDown: boolean;
  "data-testid"?: string;
  role: string;
}

const ComponentsMap: FC<ComponentsMapProps> = ({ children, ...rest }) => {
  const nonActiveCellProps = {
    onContextMenu: rest.onContextMenu,
    "data-testid": rest["data-testid"],
    role: rest.role,
  };

  switch (children) {
    case CellState.bomb:
      return (
        <BombFrame {...nonActiveCellProps}>
          <Bomb data-testid={`bomb_${rest["data-testid"]}`} />
        </BombFrame>
      );
    case CellState.hidden:
      return <ClosedFrame {...rest}>{children}</ClosedFrame>;
    case CellState.flag:
      return (
        <ClosedFrame {...rest}>
          <Flag data-testid={`flag_${rest["data-testid"]}`}>{children}</Flag>
        </ClosedFrame>
      );
    case CellState.weakFlag:
      return (
        <ClosedFrame {...rest}>
          <WeakFlag data-testid={`weakFlag_${rest["data-testid"]}`} />
        </ClosedFrame>
      );
    default:
      return <RevealedFrame {...nonActiveCellProps}>{children}</RevealedFrame>;
  }
};

interface ClosedFrameProps {
  mouseDown?: boolean;
}

export const ClosedFrame = styled.div<ClosedFrameProps>`
  display: flex;
  align-items: center;
  justify-content: center;
  user-select: none;
  cursor: pointer;
  width: 1.8vh;
  height: 1.8vh;
  color: transparent;
  background-color: #d1d1d1;
  border: 0.6vh solid transparent;
  border-color: ${({ mouseDown = false }) =>
    mouseDown ? "transparent" : "white #9e9e9e #9e9e9e white"};
  &:hover {
    filter: brightness(1.1);
  }
`;

const transparent = "rgba(0,0,0,0)";
const colors: { [key in CellType]: string } = {
  0: transparent,
  1: "#2a48ec",
  2: "#2bb13d",
  3: "#ec6561",
  4: "#233db7",
  5: "#a6070f",
  6: "#e400af",
  7: "#906a02",
  8: "#fa0707",
  9: transparent,
  10: transparent,
  11: transparent,
  12: transparent,
};

const RevealedFrame = styled(ClosedFrame)`
  border-color: #dddddd;
  cursor: default;
  color: ${({ children }) => colors[children as CellType] ?? transparent};
  &:hover {
    filter: brightness(1);
  }
`;

const Bomb = styled.div`
  border-radius: 50%;
  width: 1vh;
  height: 1vh;
  background-color: #333;
`;

const BombFrame = styled(RevealedFrame)`
  background-color: #ec433c;
`;

const Flag = styled.div`
  width: 0px;
  height: 0px;
  color: ${transparent};
  border-top: 0.5vh solid transparent;
  border-bottom: 0.5vh solid transparent;
  border-left: 0.5vh solid #ec433c;
`;

const WeakFlag = styled(Flag)`
  border-left: 0.5vh solid #f19996;
`;
