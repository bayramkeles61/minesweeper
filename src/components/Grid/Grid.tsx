import { FC } from "react";
import styled from "@emotion/styled";

import { Coords, Field } from "@/helpers/Field";

import { Cell } from "./Cell";

export interface GridProps {
  /**
   * Field data
   */
  children: Field;
  /**
   * onClick handler
   */
  onClick: (coords: Coords) => void;
  /**
   * onContextMenu handler
   */
  onContextMenu: (coords: Coords) => void;
}

export const Grid: FC<GridProps> = ({ children, ...rest }) => (
  <Wrapper size={children.length} role="grid">
    {children.map((row, y) =>
      row.map((cell, x) => (
        // Stryker disable next-line StringLiteral
        <Cell key={`${y}_${x}_${cell}`} coords={[y, x]} {...rest}>
          {cell}
        </Cell>
      )),
    )}
  </Wrapper>
);

interface WrapperProps {
  size: number;
}

const Wrapper = styled.div<WrapperProps>`
  display: grid;
  grid-template-columns: repeat(${({ size }) => size}, auto);
  width: max-content;
  padding: 1vw;
`;
