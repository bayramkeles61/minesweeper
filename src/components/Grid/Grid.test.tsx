import { it, expect } from "vitest";
import { render } from "@testing-library/react";

import { Field } from "@/helpers/Field";

import { Grid } from "./Grid";

export const MockFieldData: Field = [
  [9, 2, 9, 1, 0, 0, 1, 1, 1, 1, 1, 1],
  [1, 2, 2, 2, 1, 0, 1, 9, 1, 1, 9, 1],
  [0, 0, 1, 9, 10, 0, 2, 2, 2, 1, 1, 1],
  [0, 0, 10, 10, 1, 0, 1, 9, 1, 2, 2, 2],
  [0, 1, 1, 2, 2, 9, 1, 1, 1, 0, 0, 0],
  [0, 1, 9, 3, 9, 2, 10, 0, 0, 2, 1, 1],
  [0, 2, 2, 4, 9, 2, 10, 1, 1, 1, 9, 1],
  [0, 1, 9, 2, 1, 1, 1, 9, 1, 2, 2, 2],
  [0, 1, 10, 10, 0, 0, 1, 1, 1, 1, 9, 1],
  [0, 1, 10, 10, 0, 0, 1, 1, 1, 1, 9, 1],
  [0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 9, 1],
  [0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 9, 1],
];

it("Grid renders correctly", () => {
  const props = {
    onClick: vi.fn(),
    onContextMenu: vi.fn(),
  };

  const { asFragment } = render(<Grid {...props}>{MockFieldData}</Grid>);

  expect(asFragment()).toMatchSnapshot();
});
