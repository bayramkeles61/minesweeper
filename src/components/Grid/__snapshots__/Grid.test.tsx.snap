// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Grid renders correctly 1`] = `
<DocumentFragment>
  <div
    class="css-2gtdmb"
    role="grid"
    size="12"
  >
    <div
      class="css-12dzs03"
      data-testid="0,0"
      role="cell"
    >
      <div
        class="css-81z0jo"
        data-testid="bomb_0,0"
      />
    </div>
    <div
      class="css-13kl6wr"
      data-testid="0,1"
      role="cell"
    >
      2
    </div>
    <div
      class="css-12dzs03"
      data-testid="0,2"
      role="cell"
    >
      <div
        class="css-81z0jo"
        data-testid="bomb_0,2"
      />
    </div>
    <div
      class="css-y5m8ng"
      data-testid="0,3"
      role="cell"
    >
      1
    </div>
    <div
      class="css-1l2qare"
      data-testid="0,4"
      role="cell"
    >
      0
    </div>
    <div
      class="css-1l2qare"
      data-testid="0,5"
      role="cell"
    >
      0
    </div>
    <div
      class="css-y5m8ng"
      data-testid="0,6"
      role="cell"
    >
      1
    </div>
    <div
      class="css-y5m8ng"
      data-testid="0,7"
      role="cell"
    >
      1
    </div>
    <div
      class="css-y5m8ng"
      data-testid="0,8"
      role="cell"
    >
      1
    </div>
    <div
      class="css-y5m8ng"
      data-testid="0,9"
      role="cell"
    >
      1
    </div>
    <div
      class="css-y5m8ng"
      data-testid="0,10"
      role="cell"
    >
      1
    </div>
    <div
      class="css-y5m8ng"
      data-testid="0,11"
      role="cell"
    >
      1
    </div>
    <div
      class="css-y5m8ng"
      data-testid="1,0"
      role="cell"
    >
      1
    </div>
    <div
      class="css-13kl6wr"
      data-testid="1,1"
      role="cell"
    >
      2
    </div>
    <div
      class="css-13kl6wr"
      data-testid="1,2"
      role="cell"
    >
      2
    </div>
    <div
      class="css-13kl6wr"
      data-testid="1,3"
      role="cell"
    >
      2
    </div>
    <div
      class="css-y5m8ng"
      data-testid="1,4"
      role="cell"
    >
      1
    </div>
    <div
      class="css-1l2qare"
      data-testid="1,5"
      role="cell"
    >
      0
    </div>
    <div
      class="css-y5m8ng"
      data-testid="1,6"
      role="cell"
    >
      1
    </div>
    <div
      class="css-12dzs03"
      data-testid="1,7"
      role="cell"
    >
      <div
        class="css-81z0jo"
        data-testid="bomb_1,7"
      />
    </div>
    <div
      class="css-y5m8ng"
      data-testid="1,8"
      role="cell"
    >
      1
    </div>
    <div
      class="css-y5m8ng"
      data-testid="1,9"
      role="cell"
    >
      1
    </div>
    <div
      class="css-12dzs03"
      data-testid="1,10"
      role="cell"
    >
      <div
        class="css-81z0jo"
        data-testid="bomb_1,10"
      />
    </div>
    <div
      class="css-y5m8ng"
      data-testid="1,11"
      role="cell"
    >
      1
    </div>
    <div
      class="css-1l2qare"
      data-testid="2,0"
      role="cell"
    >
      0
    </div>
    <div
      class="css-1l2qare"
      data-testid="2,1"
      role="cell"
    >
      0
    </div>
    <div
      class="css-y5m8ng"
      data-testid="2,2"
      role="cell"
    >
      1
    </div>
    <div
      class="css-12dzs03"
      data-testid="2,3"
      role="cell"
    >
      <div
        class="css-81z0jo"
        data-testid="bomb_2,3"
      />
    </div>
    <div
      class="css-1tt7j02"
      data-testid="2,4"
      role="cell"
    >
      10
    </div>
    <div
      class="css-1l2qare"
      data-testid="2,5"
      role="cell"
    >
      0
    </div>
    <div
      class="css-13kl6wr"
      data-testid="2,6"
      role="cell"
    >
      2
    </div>
    <div
      class="css-13kl6wr"
      data-testid="2,7"
      role="cell"
    >
      2
    </div>
    <div
      class="css-13kl6wr"
      data-testid="2,8"
      role="cell"
    >
      2
    </div>
    <div
      class="css-y5m8ng"
      data-testid="2,9"
      role="cell"
    >
      1
    </div>
    <div
      class="css-y5m8ng"
      data-testid="2,10"
      role="cell"
    >
      1
    </div>
    <div
      class="css-y5m8ng"
      data-testid="2,11"
      role="cell"
    >
      1
    </div>
    <div
      class="css-1l2qare"
      data-testid="3,0"
      role="cell"
    >
      0
    </div>
    <div
      class="css-1l2qare"
      data-testid="3,1"
      role="cell"
    >
      0
    </div>
    <div
      class="css-1tt7j02"
      data-testid="3,2"
      role="cell"
    >
      10
    </div>
    <div
      class="css-1tt7j02"
      data-testid="3,3"
      role="cell"
    >
      10
    </div>
    <div
      class="css-y5m8ng"
      data-testid="3,4"
      role="cell"
    >
      1
    </div>
    <div
      class="css-1l2qare"
      data-testid="3,5"
      role="cell"
    >
      0
    </div>
    <div
      class="css-y5m8ng"
      data-testid="3,6"
      role="cell"
    >
      1
    </div>
    <div
      class="css-12dzs03"
      data-testid="3,7"
      role="cell"
    >
      <div
        class="css-81z0jo"
        data-testid="bomb_3,7"
      />
    </div>
    <div
      class="css-y5m8ng"
      data-testid="3,8"
      role="cell"
    >
      1
    </div>
    <div
      class="css-13kl6wr"
      data-testid="3,9"
      role="cell"
    >
      2
    </div>
    <div
      class="css-13kl6wr"
      data-testid="3,10"
      role="cell"
    >
      2
    </div>
    <div
      class="css-13kl6wr"
      data-testid="3,11"
      role="cell"
    >
      2
    </div>
    <div
      class="css-1l2qare"
      data-testid="4,0"
      role="cell"
    >
      0
    </div>
    <div
      class="css-y5m8ng"
      data-testid="4,1"
      role="cell"
    >
      1
    </div>
    <div
      class="css-y5m8ng"
      data-testid="4,2"
      role="cell"
    >
      1
    </div>
    <div
      class="css-13kl6wr"
      data-testid="4,3"
      role="cell"
    >
      2
    </div>
    <div
      class="css-13kl6wr"
      data-testid="4,4"
      role="cell"
    >
      2
    </div>
    <div
      class="css-12dzs03"
      data-testid="4,5"
      role="cell"
    >
      <div
        class="css-81z0jo"
        data-testid="bomb_4,5"
      />
    </div>
    <div
      class="css-y5m8ng"
      data-testid="4,6"
      role="cell"
    >
      1
    </div>
    <div
      class="css-y5m8ng"
      data-testid="4,7"
      role="cell"
    >
      1
    </div>
    <div
      class="css-y5m8ng"
      data-testid="4,8"
      role="cell"
    >
      1
    </div>
    <div
      class="css-1l2qare"
      data-testid="4,9"
      role="cell"
    >
      0
    </div>
    <div
      class="css-1l2qare"
      data-testid="4,10"
      role="cell"
    >
      0
    </div>
    <div
      class="css-1l2qare"
      data-testid="4,11"
      role="cell"
    >
      0
    </div>
    <div
      class="css-1l2qare"
      data-testid="5,0"
      role="cell"
    >
      0
    </div>
    <div
      class="css-y5m8ng"
      data-testid="5,1"
      role="cell"
    >
      1
    </div>
    <div
      class="css-12dzs03"
      data-testid="5,2"
      role="cell"
    >
      <div
        class="css-81z0jo"
        data-testid="bomb_5,2"
      />
    </div>
    <div
      class="css-1ejwm48"
      data-testid="5,3"
      role="cell"
    >
      3
    </div>
    <div
      class="css-12dzs03"
      data-testid="5,4"
      role="cell"
    >
      <div
        class="css-81z0jo"
        data-testid="bomb_5,4"
      />
    </div>
    <div
      class="css-13kl6wr"
      data-testid="5,5"
      role="cell"
    >
      2
    </div>
    <div
      class="css-1tt7j02"
      data-testid="5,6"
      role="cell"
    >
      10
    </div>
    <div
      class="css-1l2qare"
      data-testid="5,7"
      role="cell"
    >
      0
    </div>
    <div
      class="css-1l2qare"
      data-testid="5,8"
      role="cell"
    >
      0
    </div>
    <div
      class="css-13kl6wr"
      data-testid="5,9"
      role="cell"
    >
      2
    </div>
    <div
      class="css-y5m8ng"
      data-testid="5,10"
      role="cell"
    >
      1
    </div>
    <div
      class="css-y5m8ng"
      data-testid="5,11"
      role="cell"
    >
      1
    </div>
    <div
      class="css-1l2qare"
      data-testid="6,0"
      role="cell"
    >
      0
    </div>
    <div
      class="css-13kl6wr"
      data-testid="6,1"
      role="cell"
    >
      2
    </div>
    <div
      class="css-13kl6wr"
      data-testid="6,2"
      role="cell"
    >
      2
    </div>
    <div
      class="css-bqz2uz"
      data-testid="6,3"
      role="cell"
    >
      4
    </div>
    <div
      class="css-12dzs03"
      data-testid="6,4"
      role="cell"
    >
      <div
        class="css-81z0jo"
        data-testid="bomb_6,4"
      />
    </div>
    <div
      class="css-13kl6wr"
      data-testid="6,5"
      role="cell"
    >
      2
    </div>
    <div
      class="css-1tt7j02"
      data-testid="6,6"
      role="cell"
    >
      10
    </div>
    <div
      class="css-y5m8ng"
      data-testid="6,7"
      role="cell"
    >
      1
    </div>
    <div
      class="css-y5m8ng"
      data-testid="6,8"
      role="cell"
    >
      1
    </div>
    <div
      class="css-y5m8ng"
      data-testid="6,9"
      role="cell"
    >
      1
    </div>
    <div
      class="css-12dzs03"
      data-testid="6,10"
      role="cell"
    >
      <div
        class="css-81z0jo"
        data-testid="bomb_6,10"
      />
    </div>
    <div
      class="css-y5m8ng"
      data-testid="6,11"
      role="cell"
    >
      1
    </div>
    <div
      class="css-1l2qare"
      data-testid="7,0"
      role="cell"
    >
      0
    </div>
    <div
      class="css-y5m8ng"
      data-testid="7,1"
      role="cell"
    >
      1
    </div>
    <div
      class="css-12dzs03"
      data-testid="7,2"
      role="cell"
    >
      <div
        class="css-81z0jo"
        data-testid="bomb_7,2"
      />
    </div>
    <div
      class="css-13kl6wr"
      data-testid="7,3"
      role="cell"
    >
      2
    </div>
    <div
      class="css-y5m8ng"
      data-testid="7,4"
      role="cell"
    >
      1
    </div>
    <div
      class="css-y5m8ng"
      data-testid="7,5"
      role="cell"
    >
      1
    </div>
    <div
      class="css-y5m8ng"
      data-testid="7,6"
      role="cell"
    >
      1
    </div>
    <div
      class="css-12dzs03"
      data-testid="7,7"
      role="cell"
    >
      <div
        class="css-81z0jo"
        data-testid="bomb_7,7"
      />
    </div>
    <div
      class="css-y5m8ng"
      data-testid="7,8"
      role="cell"
    >
      1
    </div>
    <div
      class="css-13kl6wr"
      data-testid="7,9"
      role="cell"
    >
      2
    </div>
    <div
      class="css-13kl6wr"
      data-testid="7,10"
      role="cell"
    >
      2
    </div>
    <div
      class="css-13kl6wr"
      data-testid="7,11"
      role="cell"
    >
      2
    </div>
    <div
      class="css-1l2qare"
      data-testid="8,0"
      role="cell"
    >
      0
    </div>
    <div
      class="css-y5m8ng"
      data-testid="8,1"
      role="cell"
    >
      1
    </div>
    <div
      class="css-1tt7j02"
      data-testid="8,2"
      role="cell"
    >
      10
    </div>
    <div
      class="css-1tt7j02"
      data-testid="8,3"
      role="cell"
    >
      10
    </div>
    <div
      class="css-1l2qare"
      data-testid="8,4"
      role="cell"
    >
      0
    </div>
    <div
      class="css-1l2qare"
      data-testid="8,5"
      role="cell"
    >
      0
    </div>
    <div
      class="css-y5m8ng"
      data-testid="8,6"
      role="cell"
    >
      1
    </div>
    <div
      class="css-y5m8ng"
      data-testid="8,7"
      role="cell"
    >
      1
    </div>
    <div
      class="css-y5m8ng"
      data-testid="8,8"
      role="cell"
    >
      1
    </div>
    <div
      class="css-y5m8ng"
      data-testid="8,9"
      role="cell"
    >
      1
    </div>
    <div
      class="css-12dzs03"
      data-testid="8,10"
      role="cell"
    >
      <div
        class="css-81z0jo"
        data-testid="bomb_8,10"
      />
    </div>
    <div
      class="css-y5m8ng"
      data-testid="8,11"
      role="cell"
    >
      1
    </div>
    <div
      class="css-1l2qare"
      data-testid="9,0"
      role="cell"
    >
      0
    </div>
    <div
      class="css-y5m8ng"
      data-testid="9,1"
      role="cell"
    >
      1
    </div>
    <div
      class="css-1tt7j02"
      data-testid="9,2"
      role="cell"
    >
      10
    </div>
    <div
      class="css-1tt7j02"
      data-testid="9,3"
      role="cell"
    >
      10
    </div>
    <div
      class="css-1l2qare"
      data-testid="9,4"
      role="cell"
    >
      0
    </div>
    <div
      class="css-1l2qare"
      data-testid="9,5"
      role="cell"
    >
      0
    </div>
    <div
      class="css-y5m8ng"
      data-testid="9,6"
      role="cell"
    >
      1
    </div>
    <div
      class="css-y5m8ng"
      data-testid="9,7"
      role="cell"
    >
      1
    </div>
    <div
      class="css-y5m8ng"
      data-testid="9,8"
      role="cell"
    >
      1
    </div>
    <div
      class="css-y5m8ng"
      data-testid="9,9"
      role="cell"
    >
      1
    </div>
    <div
      class="css-12dzs03"
      data-testid="9,10"
      role="cell"
    >
      <div
        class="css-81z0jo"
        data-testid="bomb_9,10"
      />
    </div>
    <div
      class="css-y5m8ng"
      data-testid="9,11"
      role="cell"
    >
      1
    </div>
    <div
      class="css-1l2qare"
      data-testid="10,0"
      role="cell"
    >
      0
    </div>
    <div
      class="css-y5m8ng"
      data-testid="10,1"
      role="cell"
    >
      1
    </div>
    <div
      class="css-y5m8ng"
      data-testid="10,2"
      role="cell"
    >
      1
    </div>
    <div
      class="css-y5m8ng"
      data-testid="10,3"
      role="cell"
    >
      1
    </div>
    <div
      class="css-1l2qare"
      data-testid="10,4"
      role="cell"
    >
      0
    </div>
    <div
      class="css-1l2qare"
      data-testid="10,5"
      role="cell"
    >
      0
    </div>
    <div
      class="css-y5m8ng"
      data-testid="10,6"
      role="cell"
    >
      1
    </div>
    <div
      class="css-y5m8ng"
      data-testid="10,7"
      role="cell"
    >
      1
    </div>
    <div
      class="css-y5m8ng"
      data-testid="10,8"
      role="cell"
    >
      1
    </div>
    <div
      class="css-y5m8ng"
      data-testid="10,9"
      role="cell"
    >
      1
    </div>
    <div
      class="css-12dzs03"
      data-testid="10,10"
      role="cell"
    >
      <div
        class="css-81z0jo"
        data-testid="bomb_10,10"
      />
    </div>
    <div
      class="css-y5m8ng"
      data-testid="10,11"
      role="cell"
    >
      1
    </div>
    <div
      class="css-1l2qare"
      data-testid="11,0"
      role="cell"
    >
      0
    </div>
    <div
      class="css-y5m8ng"
      data-testid="11,1"
      role="cell"
    >
      1
    </div>
    <div
      class="css-y5m8ng"
      data-testid="11,2"
      role="cell"
    >
      1
    </div>
    <div
      class="css-y5m8ng"
      data-testid="11,3"
      role="cell"
    >
      1
    </div>
    <div
      class="css-1l2qare"
      data-testid="11,4"
      role="cell"
    >
      0
    </div>
    <div
      class="css-1l2qare"
      data-testid="11,5"
      role="cell"
    >
      0
    </div>
    <div
      class="css-y5m8ng"
      data-testid="11,6"
      role="cell"
    >
      1
    </div>
    <div
      class="css-y5m8ng"
      data-testid="11,7"
      role="cell"
    >
      1
    </div>
    <div
      class="css-y5m8ng"
      data-testid="11,8"
      role="cell"
    >
      1
    </div>
    <div
      class="css-y5m8ng"
      data-testid="11,9"
      role="cell"
    >
      1
    </div>
    <div
      class="css-12dzs03"
      data-testid="11,10"
      role="cell"
    >
      <div
        class="css-81z0jo"
        data-testid="bomb_11,10"
      />
    </div>
    <div
      class="css-y5m8ng"
      data-testid="11,11"
      role="cell"
    >
      1
    </div>
  </div>
</DocumentFragment>
`;
